Draft-Accepted/feedback_completer_auto_discovery.md

```diff
@@ -0,0 +1,240 @@
+
+Each item within `Feedbacks` is a folder.
+- `_startup_`: feedback providers declared in this folder will be auto-loaded at the startup of an interactive `ConsoleHost` session.
+- `git` or `kubectl`: feedback provider for a specific native command should be declared in a folder named after the native command. It will be auto-loaded when the native command gets executed.
+
+Within each sub-folder, a JSON file named after the folder name should be defined to configure the auto-discovery for the feedback provider.
+
+```JSONC
+{
+    "module": "<module-name-or-path>",  // Module to load to register the feedback provider.
+    "disable": false, // Control whether auto-discovery should find this feedback provider.
+}
+```
+
+About the `module` key:
+- Its value could be the path to an assembly DLL, in which case the DLL will be loaded as a binary module.
+- The processing order is
+  - First, expand the string. So, the value could contain PowerShell variables such as `$env:ProgramData`.
+  - Second, try resolving the value as a path using PowerShell built-in API.
+    - if it's successfully resolved as a path, use the path for module loading; Otherwise
+    - if the value contains directory separator, then discovery fails; Otherwise
+    - use the expanded value as the module name for module loading
+
+#### Execution Flow
+
+1. At startup, auto-load the enabled feedback providers from the `_startup_` folder of the `Feedbacks` paths. Do the auto-loading before processing user's profile.
+2. Report what feedback providers were processed and the time taken, in the streaming way.
+3. In `NativeCommandProcessor`, when a native command gets executed,
+   - Check if a feedback provider for this native command is available in the `Feedbacks` paths;
+   - If so, check if the specified module already loaded;
+   - If not, load it at the `DoComplete` phase, so the feedback provider will be ready right after the native command finishes.
+
+**[NOTE:]** It would be better if we can check if a feedback provider for the native command has already registered before checking the paths,
+but unlike the completer registration for native commands, today there is no mapping between a feedback provider and a native command.
+Also, it's allowed to register multiple feedback providers for a single native command.
+
+#### Discussion Points
+
+1. Should we expand the string value for `module` key, or always treat the value as literal?
+   It feels like a useful feature, but could come with security implication, especially in WDAC/AppLocker enforced environment.
+   - Note: PowerShell data file (`.psd1`) doesn't allow environment variables.
+
+2. Should we add another key to indicate the target OS?
+   - A feedback provider may only work on a specific OS, such as the `"WinGet CommandNotFound"` feedback provider only works on Windows.
+   - Such a key could be handy if a user wants to share the feedback/tab-completer configurations among multiple machines via a cloud drive.
```

Comment:
I think this could be good, thinking about the linux cmd-not-found predictor. Also if a user wants to bring this configuration across their different machines they dont need to have multiple different folder structures? I am not entirely sure how they would necessarily do it but I know some community folks use external tools to share their $PROFILE across machines.

Reply:
> they dont need to have multiple different folder structures?

Folder structures under `feedbacks` and `completions` are the same. I guess a user can create a symbolic link to make `<personal>/powershell/feedbacks` or `<personal>/powershell/completions` points to the folders from a cloud drive.